### PR TITLE
support expand-like reshape for kStitch in CPU backend

### PR DESCRIPTION
**Pull request by wyzero**

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/fusion_utils.cc [hunk 1]
```diff
@@ -1863,6 +1894,56 @@ bool StitchCPUAnalysis::doBcastOpTileAnalysis(
...
+  Value outValue = cast<lmhlo::LmhloOp>(op).getResultBuffer();
+  // target_shape should not do parallel computation.
+  if (isa<lmhlo::DynamicReshapeOp>(op)) {
+    auto& targetShape = tilePlan[op->getOperand(1)];
+    TileInfo mergedInfo = targetShape;
+    if (!mergedInfo.merge(0)) return false;
+    changed |= targetShape.updateIfNotEqual(mergedInfo);
+  }
+
+  auto propagateTileDims = [&](Value in, TileInfo& inInfo, Value out,
```
Comment (fortianyou) [inline tao_compiler/mlir/disc/transforms/fusion_utils.cc:1916]: nit: propagatableTileDims

File: tao_compiler/mlir/disc/transforms/fusion_utils.cc [hunk 2]
```diff
@@ -1863,6 +1894,56 @@ bool StitchCPUAnalysis::doBcastOpTileAnalysis(
...
+    return true;
+  };
+
+  TileInfo inMergedInfo = tilePlan[inValue];
+  TileInfo outMergedInfo = tilePlan[outValue];
+
+  if (!propagateTileDims(inValue, inMergedInfo, outValue, outMergedInfo) ||
+      !propagateTileDims(outValue, outMergedInfo, inValue, inMergedInfo)) {
+    return false;
+  }
```
Comment (fortianyou) [inline tao_compiler/mlir/disc/transforms/fusion_utils.cc:1940]: Return true only when both in->out/out->in are propagatable?

File: tao_compiler/mlir/disc/transforms/fusion_utils.cc [hunk 3]
```diff
@@ -1863,6 +1894,56 @@ bool StitchCPUAnalysis::doBcastOpTileAnalysis(
...
 
+SmallVector<int> getNonSizeOneDims(MemRefType ty) {
+  SmallVector<int> dims;
+  for (int i = 0; i < ty.getRank(); ++i)
+    if (ty.getShape()[i] != 1) dims.push_back(i);
+  return dims;
+}
+
+bool StitchCPUAnalysis::doReshapeOpTileAnalysis(
+    DenseMap<Value, TileInfo>& tilePlan, Operation* op, bool& changed) {
```
Comment (fortianyou) [inline tao_compiler/mlir/disc/transforms/fusion_utils.cc:1905]: Nit: please document what is the analysis doing?

File: tao_compiler/mlir/disc/transforms/fusion_utils.cc (hunk 2 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/fusion_utils.cc:1940]: Yes. We need the tile plans for the input and output are consistent.

File: tao_compiler/mlir/disc/transforms/fusion_utils.cc (hunk 1 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/fusion_utils.cc:1916]: using verb + noun is more common style for function name? Furthermore, this function actually also update the passed inputs.

**Review by fortianyou — APPROVED** [review]
LGTM

**Pull request merged**